lib/Sema/MiscDiagnostics.cpp

```diff
@@ -2589,7 +2590,69 @@ class VarDeclUsageChecker : public ASTWalker {
     return { true, S };
   }
 };
+
+/// An AST walker that discards diagnostics for variable declarations found
+/// in inactive if-config clauses.
+class InactiveVarDeclUsageChecker : public ASTWalker {
+  DeclContext *DC;
+  
+public:
+  InactiveVarDeclUsageChecker(DeclContext *DC) : DC(DC) {}
+  
+  bool walkToDeclPre(Decl *D) override {
+    if (isa<TypeDecl>(D))
+      return false;
+    
+    // The body of #if clauses are not walked into, we need custom processing
+    // for them.
+    if (auto *ICD = dyn_cast<IfConfigDecl>(D))
+      handleIfConfig(ICD);
+    
+    return true;
+  }
   
+  /// handle #if directives. All of the active clauses will generate
+  /// diagnostics, but we also want to handle the inactive ones to avoid false
+  /// positives.
```

Comment:
I left the behavior as exists currently in Swift, and I do recall a forums discussion about this long before I began this project. I'll pull up a link if I can find it. You can also read @xwu 's [comment here](https://github.com/apple/swift/pull/36134#issuecomment-786629052).

The conclusion from the discussion was that we don't want to produce different diagnostics while compiling in different environments, so we ignore everything within #if declarations.

Because a diagnostic feature like this does not exist currently, I'm leaving the existing behavior -- but we can decide to alter it in the future.